docs/design/2021-08-18-charsets.md

```diff
@@ -0,0 +1,270 @@
+-- mysql 61cee5 
+
+// 0xe4b880 is an invalid gbk hex literal but valid utf8mb4 hex literal.
+select hex(concat(a, 0xe4b880)) from t3;
+-- mysql 61e4b880 (test on mysql 5.7 and 8.0.22)
+-- mysql returns "Cannot convert string '\x80' from binary to gbk" (test on mysql 8.0.25 and 8.0.26). TiDB will be compatible with this behavior.
+
+// 0x80 is a hex literal that invalid for neither gbk nor utf8mb4.
+select hex(concat(a, 0x80)) from t3;
+-- mysql 6180 (test on mysql 5.7 and 8.0.22)
+-- mysql returns "Cannot convert string '\x80' from binary to gbk" (test on mysql 8.0.25 and 8.0.26). TiDB will be compatible with this behavior.
+
+set @@sql_mode = '';
+insert into t3 values (0x80);
+-- mysql gets a warning and insert null values (warning: "Incorrect string value: '\x80' for column 'a' at row 1")
+
+set @@sql_mode = 'STRICT_TRANS_TABLES';
+insert into t3 values (0x80);
+-- mysql returns "Incorrect string value: '\x80' for column 'a' at row 1"
+```
+
+#### Compatibility with other components
+
+After using this version of TiDB, and when gbk-encoded data is already stored in the database, you can only use the components corresponding to this version or greater.
+- TiKV
+  - Coprocessor related builtin functions need to be processed, and collation related functions need to be implemented.
+- TiCDC
+  - It needs to be converted to gbk character set encoding when restoring the transaction downstream or outputting in accordance with the [TiCDC Open Protocol](https://docs.pingcap.com/zh/tidb/dev/ticdc-open-protocol).
+- TiFlash
+  - Related builtin functions need to be processed, and collation related functions need to be implemented.
+- BR
+  - BR realizes backup and recovery by directly manipulating SST files. At present, BR only rewrites the table ID and index ID in the key when restoring, and needs to decode KV. The decoder reuses the codec interface of TiDB, and only needs to update the TiDB dependency. There are no compatibility issues.
+- Lightning
+  - tidb-lightning is responsible for parsing SQL to become a KV pair. It implements its own parse, and special processing may be required here.
+- DM
+  - Currently, DM will specify utf8mb4 when connecting to TiDB, which needs to be processed here.
+  - When DDL is synchronized, TiDB-parser will be used to parse the DDL statement. Currently, empty charset and collation are used. Use the charset in binlog to pass in the Parser function (it is expected to be available in the [header of the DDL event](https://dev.mysql.com/doc/internals/en/query-event.html#q-charset-database-code)):
+    - If it supports (ascii, binary, latin1, utf-8, utf8mb4, gbk), handle it directly.
+    - If it is not supported, fallback to an empty charset/collation to keep the old behavior.
+- Dumpling
+  - Dumpling obtains the table structure and data of upstream TiDB/MySQL by executing SQL statements, so there is no compatibility problem.
+- TiDB-binlog
+  - Drainer needs to convert to gbk encoding when outputting data downstream.
```

Comment:
(ditto with TiCDC)

Reply:
I think we can make sure drainer writes data with the connection charset setting to `UTF8`.
If users want GBK data, they can set their own connection charset to `GBK`. In this way we needn't convert data in drainer. (Binlogs from TiDB are assured to be `UTF` format)

Reply:
Got it.
So now TiDB writes data to binlog in UTF8 encoding, this component does not need to change.